File: core/plugin/__DEV/Python/n_gfDebugVector.py
# -*- coding: utf-8 -*-
"""
Copyright (c) 2019 Giuliano França

Redistribution:
    Something here.

Maya Node:
    [This is a prototype version of the gfDebugVector node. You should be using the related C++ version.]
    This node is a test node and only performs test operations with one input value.

Requirements:
    Maya 2017 or above.

Todo:
    * Lock the SRT controls in post-constructor
    * https://en.wikibooks.org/wiki/OpenGL_Programming/GLStart/Tut3
    * https://community.khronos.org/t/fyi-c-code-for-drawing-a-line-3d-arrow-head/70602
    * glFT.glEnable(omr1.MGL_BLEND); glFT.glBlendFunc(omr1.MGL_SRC_ALPHA, omr1.MGL_ONE_MINUS_SRC_ALPHA)
    * Fix can't select in xray mode

This code supports Pylint. Rc file in project.
"""
# pylint: disable=import-error
# import-error = Supress Maya modules import error

import maya.api.OpenMaya as om2
import maya.api.OpenMayaUI as omui2
import maya.api.OpenMayaRender as omr2


def maya_useNewAPI():
    """ Function to Maya recognize the use of the Python API 2.0. """
    # pylint: disable=invalid-name, unnecessary-pass
    pass


def INPUT_ATTR(FNATTR):
    """ Configure a input attribute. """
    # pylint: disable=invalid-name
    FNATTR.writable = True
    FNATTR.readable = True
    FNATTR.storable = True
    FNATTR.keyable = True


def OUTPUT_ATTR(FNATTR):
    """ Configure a output attribute. """
    # pylint: disable=invalid-name
    FNATTR.writable = False
    FNATTR.readable = True
    FNATTR.storable = False
    FNATTR.keyable = False


class DebugVector(omui2.MPxLocatorNode):
    """ Main class of gfDebugVector node. """

    kNODE_NAME = ""
    kNODE_CLASSIFY = ""
    kNODE_REGISTRANT_ID = ""
    kNODE_ID = ""

    inLineWidth = om2.MObject()
    inColor = om2.MObject()
    inRadius = om2.MObject()
    inXRay = om2.MObject()
    inVec1 = om2.MObject()
    inVec2 = om2.MObject()
    inOperation = om2.MObject()

    inSize = om2.MObject()

    def __init__(self):
        """ Constructor. """
        omui2.MPxLocatorNode.__init__(self)

    def postConstructor(self):
        """ Post Constructor. """
        thisMob = self.thisMObject()
        om2.MFnDependencyNode(thisMob).setName("gfDebugVector_PShape#")

    @staticmethod
    def creator():
        """ Maya creator function. """
        return DebugVector()

    @staticmethod
    def initialize():
        """
        Defines the set of attributes for this node. The attributes declared in this function are assigned
        as static members to DebugVector class. Instances of DebugVector will use these attributes to create plugs
        for use in the compute() method.
        """
        eAttr = om2.MFnEnumAttribute()
        nAttr = om2.MFnNumericAttribute()

        DebugVector.inLineWidth = nAttr.create("lineWidth", "lw", om2.MFnNumericData.kFloat, 3.0)
        nAttr.setMin(1.0)
        nAttr.setSoftMax(5.0)
        INPUT_ATTR(nAttr)

        DebugVector.inColor = nAttr.createColor("color", "color")
        nAttr.default = (1.0, 1.0, 0.0)
        INPUT_ATTR(nAttr)

        DebugVector.inRadius = nAttr.create("radius", "radius", om2.MFnNumericData.kFloat, 1.0)
        nAttr.setMin(0.0)
        nAttr.setSoftMax(2.0)
        INPUT_ATTR(nAttr)

        DebugVector.inXRay = nAttr.create("XRay", "xray", om2.MFnNumericData.kBoolean, False)
        INPUT_ATTR(nAttr)

        DebugVector.inOperation = eAttr.create("operation", "op", 1)
        eAttr.addField("No operation", 0)
        eAttr.addField("Add", 1)
        eAttr.addField("Subtract", 2)
        INPUT_ATTR(eAttr)

        DebugVector.inVec1 = nAttr.createPoint("vector1", "v1")
        INPUT_ATTR(nAttr)

        DebugVector.inVec2 = nAttr.createPoint("vector2", "v2")
        INPUT_ATTR(nAttr)

        DebugVector.addAttribute(DebugVector.inLineWidth)
        DebugVector.addAttribute(DebugVector.inColor)
        DebugVector.addAttribute(DebugVector.inRadius)
        DebugVector.addAttribute(DebugVector.inXRay)
        DebugVector.addAttribute(DebugVector.inOperation)
        DebugVector.addAttribute(DebugVector.inVec1)
        DebugVector.addAttribute(DebugVector.inVec2)

    def compute(self, plug, dataBlock):
        """
        Node computation method:
            * plug is a connection point related to one of our node attributes (either an input or an output).
            * dataBlock contains the data on which we will base our computations.
        """
        # pylint: disable=no-self-use, unused-argument
        return None

    def draw(self, view, path, style, status):
        """
        Draw custom geometry in the viewport using OpenGL calls.
            * view [M3dView] is a 3D view that is being drawn into.
            * path [MDagPath] to the parent (transform node) of this locator in the DAG.  To obtain the locator shape node,
                use MDagPath::extendToShape() if there is only one shape node under the transform or
                MDagPath::extendToShapeDirectlyBelow(unsigned int index) with the shape index if there are multiple
                shapes under the transform.
            * style [M3dView.DisplayStyle] is the style to draw object in.
            * status [M3dView.DisplayStatus] is the selection status of the object.
        """
        # pylint: disable=unused-argument
        thisMob = self.thisMObject()
        lineW = om2.MPlug(thisMob, DebugVector.inLineWidth).asFloat()
        colorLoc = om2.MPlug(thisMob, DebugVector.inColor).asMDataHandle().asFloatVector()
        # radius = om2.MPlug(thisMob, DebugVector.inRadius).asFloat()
        xray = om2.MPlug(thisMob, DebugVector.inXRay).asBool()
        vVec1 = om2.MPlug(thisMob, DebugVector.inVec1).asMDataHandle().asFloatVector()
        vVec2 = om2.MPlug(thisMob, DebugVector.inVec2).asMDataHandle().asFloatVector()

        view.beginGL()

        # Drawing in VP1 views will be done using Python API 1.0
        import maya.OpenMayaRender as omr1
        glRenderer = omr1.MHardwareRenderer.theRenderer()
        glFT = glRenderer.glFunctionTable()

        glFT.glPushAttrib(omr1.MGL_CURRENT_BIT)
        glFT.glDisable(omr1.MGL_CULL_FACE)
        if xray:
            glFT.glEnable(omr1.MGL_DEPTH_TEST)
            glFT.glClear(omr1.MGL_DEPTH_BUFFER_BIT)

        if status == omui2.M3dView.kActive:
            glFT.glColor3f(0.3, 1.0, 1.0)
        elif status == omui2.M3dView.kLead:
            glFT.glColor3f(1.0, 1.0, 1.0)
        elif status == omui2.M3dView.kDormant:
            glFT.glColor3f(colorLoc.x, colorLoc.y, colorLoc.z)

        glFT.glLineWidth(lineW)
        glFT.glBegin(omr1.MGL_LINES)
        glFT.glVertex3f(vVec1.x, vVec1.y, vVec1.z)
        glFT.glVertex3f(vVec2.x, vVec2.y, vVec2.z)
        glFT.glEnd()
        if xray:
            glFT.glDisable(omr1.MGL_DEPTH_TEST)
        glFT.glPopAttrib()
        glFT.glLineWidth(1.0)

        view.endGL()


class DebugVectorData(om2.MUserData):
    """ Custom user data class. """

    def __init__(self):
        """ Constructor. """
        om2.MUserData.__init__(self, False)  # Don't delete after draw

        self.fDormantColor = om2.MColor()
        self.fActiveColor = om2.MColor()
        self.fLeadColor = om2.MColor()
        self.fLineList = om2.MPointArray()
        self.fLineWidth = 1.0
        self.fXRay = 5


class DebugVectorDrawOverride(omr2.MPxDrawOverride):
    """ Drawing override to enable to draw in VP2.0. """

    def __init__(self, obj):
        """ Constructor. """
        omr2.MPxDrawOverride.__init__(self, obj, DebugVectorDrawOverride.draw)

        # We want to perform custom bounding box drawing so return True so that the
        # internal rendering code will not draw it for us.
        self.mCustomBoxDraw = True
        self.mCurrentBoundingBox = om2.MBoundingBox()

    def supportedDrawAPIs(self):
        """ Select which draw API works. """
        return omr2.MRenderer.kOpenGL | omr2.MRenderer.kDirectX11 | omr2.MRenderer.kOpenGLCoreProfile

    @staticmethod
    def creator(obj):
        """ Maya creator function. """
        return DebugVectorDrawOverride(obj)

    @staticmethod
    def draw(context, data):
        """ Draw method. """
        # pylint: disable=unused-argument
        return

    def isBounded(self, objPath, cameraPath):
        """ isBounded? """
        # pylint: disable=unused-argument
        return True

    def boundingBox(self, objPath, cameraPath):
        """ Return the boundingBox """
        # pylint: disable=unused-argument
        node = objPath.node()
        vVec1 = om2.MPlug(node, DebugVector.inVec1).asMDataHandle().asFloatVector()
        vVec2 = om2.MPlug(node, DebugVector.inVec2).asMDataHandle().asFloatVector()
        corner1 = om2.MPoint(vVec1.x, vVec1.y, vVec1.z)
        corner2 = om2.MPoint(vVec2.x, vVec2.y, vVec2.z)

        self.mCurrentBoundingBox.clear()
        self.mCurrentBoundingBox.expand(corner1)
        self.mCurrentBoundingBox.expand(corner2)

        return self.mCurrentBoundingBox

    def disableInternalBoundingBoxDraw(self):
        """ Disable internal bounding box draw. """
        return self.mCustomBoxDraw

    def prepareForDraw(self, objPath, cameraPath, frameContext, oldData):
        """
        Called by Maya each time the object needs to be drawn.
        Any data needed from the Maya dependency graph must be retrieved and cached in this stage.
        Returns the data to be passed to the draw callback method.
            * objPath [MDagPath] is the path to the object being drawn.
            * cameraPath [MDagPath] is the path to the camera that is being used to draw.
            * frameContext [MFrameContext] is the frame level context information.
            * oldData [MUserData] is the data cached by the previous draw of the instance.
        """
        # pylint: disable=unused-argument
        data = oldData
        if not isinstance(data, DebugVectorData):
            data = DebugVectorData()

        node = objPath.node()
        lineW = om2.MPlug(node, DebugVector.inLineWidth).asFloat()
        colorLoc = om2.MPlug(node, DebugVector.inColor).asMDataHandle().asFloatVector()
        xray = om2.MPlug(node, DebugVector.inXRay).asBool()
        vVec1 = om2.MPlug(node, DebugVector.inVec1).asMDataHandle().asFloatVector()
        vVec2 = om2.MPlug(node, DebugVector.inVec2).asMDataHandle().asFloatVector()

        data.fLineList.clear()
        data.fLineList.append(om2.MPoint(vVec1.x, vVec1.y, vVec1.z))
        data.fLineList.append(om2.MPoint(vVec2.x, vVec2.y, vVec2.z))
        data.fDormantColor = om2.MColor([colorLoc.x, colorLoc.y, colorLoc.z])
        data.fActiveColor = om2.MColor([0.3, 1.0, 1.0])
        data.fLeadColor = om2.MColor([1.0, 1.0, 1.0])
        data.fLineWidth = lineW
        if xray:
            data.fXRay = True
        else:
            data.fXRay = False

        return data

    def hasUIDrawables(self):
        """ Has ui drawables? """
        return True

    def addUIDrawables(self, objPath, drawManager, frameContext, data):
        """
        Provides access to the MUIDrawManager, which can be used to queue up operations to draw simple UI
        shapes like lines, circles, text, etc.
        It is called after prepareForDraw() and carries the same restrictions on the sorts of operations it
        can perform.
            * objPath [MDagPath] is the path to the object being drawn.
            * drawManager [MUIDrawManager] it can be used to draw some simple geometry including text.
            * frameContext [MFrameContext] is the frame level context information.
            * data [MUserData] is the data cached by the prepareForDraw().
        """
        # pylint: disable=unused-argument
        locatorData = data
        if not isinstance(locatorData, DebugVectorData):
            return

        # status = omui2.M3dView.displayStatus(objPath)
        status = omr2.MGeometryUtilities.displayStatus(objPath)

        if locatorData.fXRay:
            drawManager.beginDrawInXray()
        else:
            drawManager.beginDrawable()

        if status == omr2.MGeometryUtilities.kActive:
            drawManager.setColor(locatorData.fActiveColor)
        elif status == omr2.MGeometryUtilities.kLead:
            drawManager.setColor(locatorData.fLeadColor)
        elif status == omr2.MGeometryUtilities.kDormant:
            drawManager.setColor(locatorData.fDormantColor)

        drawManager.setDepthPriority(5)

        drawManager.setLineWidth(locatorData.fLineWidth)
        drawManager.mesh(omr2.MUIDrawManager.kLines, locatorData.fLineList)

        if locatorData.fXRay:
            drawManager.endDrawInXray()
        else:
            drawManager.endDrawable()

File: core/plugin/__DEV/C++/source/n_gfDebugVector.cpp
#include "n_gfDebugVector.h"

// Configure a input attribute.
#define INPUT_ATTR(FNATTR)		\
	FNATTR.setWritable(true);	\
	FNATTR.setReadable(true);	\
	FNATTR.setStorable(true);	\
	FNATTR.setKeyable(true);	\

// Configure a output attribute.
#define OUTPUT_ATTR(FNATTR)		\
	FNATTR.setWritable(false);	\
	FNATTR.setReadable(true);	\
	FNATTR.setStorable(false);	\
	FNATTR.setKeyable(false);	\


// Constructor.
DebugVector::DebugVector() {}

// Destructor.
DebugVector::~DebugVector() {}

MObject DebugVector::inLineWidth;
MObject DebugVector::inColor;
MObject DebugVector::inRadius;
MObject DebugVector::inTipSize;
MObject DebugVector::inSubdivisions;
MObject DebugVector::inXRay;
MObject DebugVector::inOperation;
MObject DebugVector::inVec1;
MObject DebugVector::inVec2;
MObject DebugVector::inNormalize;
MObject DebugVector::outVector;
const MColor DebugVector::kActiveColor = MColor(0.3f, 1.0f, 1.0f);
const MColor DebugVector::kLeadColor = MColor(1.0f, 1.0f, 1.0f);

void DebugVector::postConstructor(){
    // Post Constructor.
    MObject thisMob = thisMObject();
    MFnDependencyNode(thisMob).setName("gfDebugVectorShape#");
}


void* DebugVector::creator(){
    // Maya creator function.
    return new DebugVector();
}

MStatus DebugVector::initialize(){
    /*
    Defines the set of attributes for this node. The attributes declared in this function are assigned
    as static members to DebugVector class. Instances of DebugVector will use these attributes to create plugs
    for use in the compute() method.
    */
    MStatus status;
    MFnNumericAttribute nAttr;
    MFnEnumAttribute eAttr;

    inLineWidth = nAttr.create("lineWidth", "lw", MFnNumericData::kFloat, 3.0f, &status);
    nAttr.setMin(1.0f);
    nAttr.setSoftMax(5.0f);
    INPUT_ATTR(nAttr);

    inColor = nAttr.createColor("color", "color", &status);
    nAttr.setDefault(1.0, 1.0, 0.0);
    INPUT_ATTR(nAttr);

    inTipSize = nAttr.create("tipSize", "tipSize", MFnNumericData::kFloat, 0.1f, &status);
    nAttr.setMin(0.1f);
    nAttr.setMax(1.0f);
    INPUT_ATTR(nAttr);

    inSubdivisions = nAttr.create("subdivisions", "subd", MFnNumericData::kInt, 4, &status);
    nAttr.setMin(2);
    nAttr.setMax(12);
    INPUT_ATTR(nAttr);

    inRadius = nAttr.create("radius", "radius", MFnNumericData::kFloat, 1.0f, &status);
    nAttr.setMin(0.0f);
    nAttr.setSoftMax(5.0f);
    INPUT_ATTR(nAttr);

    inXRay = nAttr.create("XRay", "XRay", MFnNumericData::kBoolean, false, &status);
    INPUT_ATTR(nAttr);

    inOperation = eAttr.create("operation", "op", 0, &status);
    eAttr.addField("No Operation", 0);
    eAttr.addField("Add", 1);
    eAttr.addField("Subtract", 2);
    eAttr.addField("Cross Product", 3);
    INPUT_ATTR(eAttr);

    inVec1 = nAttr.createPoint("vector1", "v1", &status);
    INPUT_ATTR(nAttr);

    inVec2 = nAttr.createPoint("vector2", "v2", &status);
    INPUT_ATTR(nAttr);

    inNormalize = nAttr.create("normalizeOutput", "no", MFnNumericData::kBoolean, false, &status);
    INPUT_ATTR(nAttr);

    outVector = nAttr.createPoint("outVector", "ov", &status);
    OUTPUT_ATTR(nAttr);

    addAttribute(inLineWidth);
    addAttribute(inColor);
    addAttribute(inTipSize);
    addAttribute(inSubdivisions);
    addAttribute(inRadius);
    addAttribute(inXRay);
    addAttribute(inOperation);
    addAttribute(inVec1);
    addAttribute(inVec2);
    addAttribute(inNormalize);
    addAttribute(outVector);
    attributeAffects(inOperation, outVector);
    attributeAffects(inVec1, outVector);
    attributeAffects(inVec2, outVector);
    attributeAffects(inNormalize, outVector);

    return status;
}

void DebugVector::drawArrow(MFloatVector& startPnt, MFloatVector& endPnt, float size,
                            float radius, int subd, float lineW){
    // Drawing arrow in viewport 1.0
    float tipSize = 1.0f - size;
    double step = 2.0 * M_PI / subd;
    MFloatVector vAim = endPnt - startPnt;
    MFloatVector vBaseOrigin = vAim * tipSize;
    MFloatVector nAim = vAim.normal();
    MFloatVector nWorld(0.0f, 1.0f, 0.0f);
    MFloatVector nBinormal = nWorld ^ nAim;
    nBinormal.normalize();
    MFloatVector nNormal = nAim ^ nBinormal;
    nNormal.normalize();
    double aim[4][4] = {
        nAim.x, nAim.y, nAim.z, 0.0,
        nNormal.x, nNormal.y, nNormal.z, 0.0,
        nBinormal.y, nBinormal.y, nBinormal.z, 0.0,
        startPnt.x, startPnt.y, startPnt.z, 1.0
    };
    MMatrix mBase = MMatrix(aim);
    MMatrix mOrigin = MMatrix();
    mOrigin[3][0] = vBaseOrigin.length();
    MMatrix mBaseOrigin = mOrigin * mBase;
    glLineWidth(lineW);
    glBegin(GL_LINES);
    glVertex3f(startPnt.x, startPnt.y, startPnt.z);
    glVertex3f(endPnt.x, endPnt.y, endPnt.z);
    glEnd();
    for (int i = 0; i < subd; i++){
        double theta = step * i;
        MMatrix mPoint = MMatrix();
        mPoint[3][1] = std::cos(theta) * radius;
        mPoint[3][2] = std::sin(theta) * radius;
        MMatrix mArrow = mPoint * mBaseOrigin;
        glBegin(GL_LINES);
        glVertex3d(mBaseOrigin[3][0], mBaseOrigin[3][1], mBaseOrigin[3][2]);
        glVertex3d(mArrow[3][0], mArrow[3][1], mArrow[3][2]);
        glVertex3d(mArrow[3][0], mArrow[3][1], mArrow[3][2]);
        glVertex3f(endPnt.x, endPnt.y, endPnt.z);
        glEnd();
    }
}

void DebugVector::drawArrow(MFloatVector& startPnt, MFloatVector& endPnt, float size, 
                            float radius, int subd, MPointArray& lineList){
    // Drawing arrow in viewport 2.0
    float tipSize = 1.0f - size;
    double step = 2.0 * M_PI / subd;
    MFloatVector vAim = endPnt - startPnt;
    MFloatVector vBaseOrigin = vAim * tipSize;
    MFloatVector nAim = vAim.normal();
    MFloatVector nWorld(0.0f, 1.0f, 0.0f);
    MFloatVector nBinormal = nWorld ^ nAim;
    nBinormal.normalize();
    MFloatVector nNormal = nAim ^ nBinormal;
    nNormal.normalize();
    double aim[4][4] = {
        nAim.x, nAim.y, nAim.z, 0.0,
        nNormal.x, nNormal.y, nNormal.z, 0.0,
        nBinormal.y, nBinormal.y, nBinormal.z, 0.0,
        startPnt.x, startPnt.y, startPnt.z, 1.0
    };
    MMatrix mBase = MMatrix(aim);
    MMatrix mOrigin = MMatrix();
    mOrigin[3][0] = vBaseOrigin.length();
    MMatrix mBaseOrigin = mOrigin * mBase;
    lineList.append(MPoint(startPnt));
    lineList.append(MPoint(endPnt));
    for (int i = 0; i < subd; i++){
        double theta = step * i;
        MMatrix mPoint = MMatrix();
        mPoint[3][1] = std::cos(theta) * radius;
        mPoint[3][2] = std::sin(theta) * radius;
        MMatrix mArrow = mPoint * mBaseOrigin;
        lineList.append(MPoint(mBaseOrigin[3][0], mBaseOrigin[3][1], mBaseOrigin[3][2]));
        lineList.append(MPoint(mArrow[3][0], mArrow[3][1], mArrow[3][2]));
        lineList.append(MPoint(mArrow[3][0], mArrow[3][1], mArrow[3][2]));
        lineList.append(MPoint(endPnt));
    }
}

MStatus DebugVector::compute(const MPlug& plug, MDataBlock& dataBlock){
    /*
    Node computation method:
        * plug is a connection point related to one of our node attributes (either an input or an output).
        * dataBlock contains the data on which we will base our computations.
    */
    if (plug != outVector)
        return MStatus::kUnknownParameter;

    short operation = dataBlock.inputValue(inOperation).asShort();
    MFloatVector vVector1 = dataBlock.inputValue(inVec1).asFloatVector();
    MFloatVector vVector2 = dataBlock.inputValue(inVec2).asFloatVector();
    bool normalize = dataBlock.inputValue(inNormalize).asBool();

    MFloatVector vStart, vEnd, vFinal;
    switch (operation)
    {
    case 0:
        vEnd = vVector1;
        break;
    case 1:
        vFinal = vVector1 + vVector2;
        vEnd = vFinal;
        break;
    case 2:
        vFinal = vVector1 - vVector2;
        vEnd = vFinal;
        break;
    case 3:
        vFinal = vVector1 ^ vVector2;
        vEnd = vFinal;
        break;
    }

    vStart = MFloatVector();
    if (normalize)
        vEnd.normalize();

    MDataHandle outVectorHandle = dataBlock.outputValue(outVector);
    outVectorHandle.setMFloatVector(vEnd);
    outVectorHandle.setClean();

    return MStatus::kSuccess;
}

void DebugVector::draw(M3dView& view, const MDagPath& path, M3dView::DisplayStyle style,
                       M3dView::DisplayStatus status){
    /*
    Draw custom geometry in the viewport using OpenGL calls.
        * view [M3dView] is a 3D view that is being drawn into.
        * path [MDagPath] to the parent (transform node) of this locator in the DAG.  To obtain the locator shape node,
            use MDagPath::extendToShape() if there is only one shape node under the transform or
            MDagPath::extendToShapeDirectlyBelow(unsigned int index) with the shape index if there are multiple
            shapes under the transform.
        * style [M3dView.DisplayStyle] is the style to draw object in.
        * status [M3dView.DisplayStatus] is the selection status of the object.
    */
    MObject thisMob = thisMObject();
    float lineW = MPlug(thisMob, inLineWidth).asFloat();
    MFloatVector color = MPlug(thisMob, inColor).asMDataHandle().asFloatVector();
    float tipSize = MPlug(thisMob, inTipSize).asFloat();
    int subd = MPlug(thisMob, inSubdivisions).asInt();
    float radius = MPlug(thisMob, inRadius).asFloat();
    bool xray = MPlug(thisMob, inXRay).asBool();

    short operation = MPlug(thisMob, inOperation).asShort();
    MFloatVector vVector1 = MPlug(thisMob, inVec1).asMDataHandle().asFloatVector();
    MFloatVector vVector2 = MPlug(thisMob, inVec2).asMDataHandle().asFloatVector();
    bool normalize = MPlug(thisMob, inNormalize).asBool();

    MFloatVector vStart, vEnd, vFinal;
    switch (operation)
    {
    case 0:
        vEnd = vVector1;
        break;
    case 1:
        vFinal = vVector1 + vVector2;
        vEnd = vFinal;
        break;
    case 2:
        vFinal = vVector1 - vVector2;
        vEnd = vFinal;
        break;
    case 3:
        vFinal = vVector1 ^ vVector2;
        vEnd = vFinal;
        break;
    }

    vStart = MFloatVector();
    if (normalize)
        vEnd.normalize();

    view.beginGL();

    glPushAttrib(GL_CURRENT_BIT);
    glDisable(GL_CULL_FACE);
    if (xray){
        glEnable(GL_DEPTH_TEST);
        glClear(GL_DEPTH_BUFFER_BIT);
    }

    if (status == M3dView::kActive)
        glColor3f(kActiveColor.r, kActiveColor.g, kActiveColor.b);
    else if (status == M3dView::kLead)
        glColor3f(kLeadColor.r, kLeadColor.g, kLeadColor.b);
    else if (status == M3dView::kDormant)
        glColor3f(color.x, color.y, color.z);

    DebugVector::drawArrow(vStart, vEnd, tipSize, radius, subd, lineW);

    if (xray)
        glDisable(GL_DEPTH_TEST);
    glPopAttrib();
    glLineWidth(1.0f);

    view.endGL();
}
//---------------------------------------------------------------------------------
//---------------------------------------------------------------------------------
//----------   Viewport 2.0 Implementation   --------------------------------------
//---------------------------------------------------------------------------------
//---------------------------------------------------------------------------------

DebugVectorDrawOverride::DebugVectorDrawOverride(const MObject& obj)
                       : MHWRender::MPxDrawOverride(obj, NULL, false)
                       , fDebugVector(obj){
    mCurrentBoundingBox = MBoundingBox();
    fModelEditorChangedCbId = MEventMessage::addEventCallback(
        "modelEditorChanged", OnModelEditorChanged, this);

    MGlobal::displayInfo("Constructed!");
}

DebugVectorDrawOverride::~DebugVectorDrawOverride() {
    if (fModelEditorChangedCbId != 0){
        MMessage::removeCallback(fModelEditorChangedCbId);
        fModelEditorChangedCbId = 0;
    }
}

void DebugVectorDrawOverride::OnModelEditorChanged(void* clientData){
    // Mark the node as being dirty so that it can update on display appearance
    // switch among wireframe and shaded.
    DebugVectorDrawOverride *ovr = static_cast<DebugVectorDrawOverride*>(clientData);
    if (ovr)
        MHWRender::MRenderer::setGeometryDrawDirty(ovr->fDebugVector);
}

MHWRender::MPxDrawOverride* DebugVectorDrawOverride::creator(const MObject& obj){
    // MPxDrawOverride creator function.
    return new DebugVectorDrawOverride(obj);
}

bool DebugVectorDrawOverride::isBounded(const MDagPath& objPath, 
                                        const MDagPath& cameraPath){
    // isBounded?
    return true;
}

MBoundingBox DebugVectorDrawOverride::boundingBox(const MDagPath& objPath,
                                                  const MDagPath& cameraPath){
    // Return the bounding box
    MObject node = objPath.node();
    short operation = MPlug(node, DebugVector::inOperation).asShort();
    MFloatVector vVector1 = MPlug(node, DebugVector::inVec1).asMDataHandle().asFloatVector();
    MFloatVector vVector2 = MPlug(node, DebugVector::inVec2).asMDataHandle().asFloatVector();
    bool normalize = MPlug(node, DebugVector::inNormalize).asBool();

    MFloatVector vStart, vEnd, vFinal;
    switch (operation)
    {
    case 0:
        vEnd = vVector1;
        break;
    case 1:
        vFinal = vVector1 + vVector2;
        vEnd = vFinal;
        break;
    case 2:
        vFinal = vVector1 - vVector2;
        vEnd = vFinal;
        break;
    case 3:
        vFinal = vVector1 ^ vVector2;
        vEnd = vFinal;
        break;
    }

    vStart = MFloatVector();
    if (normalize)
        vEnd.normalize();

    MPoint corner1 = MPoint(vStart.x, vStart.y, vStart.z);
    MPoint corner2 = MPoint(vEnd.x, vEnd.y, vEnd.z);
    this->mCurrentBoundingBox.clear();                // this->
    this->mCurrentBoundingBox.expand(corner1);
    this->mCurrentBoundingBox.expand(corner2);

    return this->mCurrentBoundingBox;
}

MUserData* DebugVectorDrawOverride::prepareForDraw(const MDagPath& objPath, 
                                                   const MDagPath& cameraPath,
                                                   const MHWRender::MFrameContext& frameContext,
                                                   MUserData* oldData){
    /*
    Called by Maya each time the object needs to be drawn.
    Any data needed from the Maya dependency graph must be retrieved and cached in this stage.
    Returns the data to be passed to the draw callback method.
        * objPath [MDagPath] is the path to the object being drawn.
        * cameraPath [MDagPath] is the path to the camera that is being used to draw.
        * frameContext [MFrameContext] is the frame level context information.
        * oldData [MUserData] is the data cached by the previous draw of the instance.
    */
    MGlobal::displayInfo("Getting ready to draw some cool stuff");
    DebugVectorData* data = dynamic_cast<DebugVectorData*>(oldData);
    if (!data)
        data = new DebugVectorData();

    MObject node = objPath.node();
    float lineW = MPlug(node, DebugVector::inLineWidth).asFloat();
    MFloatVector color = MPlug(node, DebugVector::inColor).asMDataHandle().asFloatVector();
    float tipSize = MPlug(node, DebugVector::inTipSize).asFloat();
    int subd = MPlug(node, DebugVector::inSubdivisions).asInt();
    float radius = MPlug(node, DebugVector::inRadius).asFloat();
    bool xray = MPlug(node, DebugVector::inXRay).asBool();
    short operation = MPlug(node, DebugVector::inOperation).asShort();
    MFloatVector vVector1 = MPlug(node, DebugVector::inVec1).asMDataHandle().asFloatVector();
    MFloatVector vVector2 = MPlug(node, DebugVector::inVec2).asMDataHandle().asFloatVector();
    bool normalize = MPlug(node, DebugVector::inNormalize).asBool();

    data->fDormantColor = MColor(color.x, color.y, color.z);
    data->fActiveColor = DebugVector::kActiveColor;
    data->fLeadColor = DebugVector::kLeadColor;
    data->fLineWidth = lineW;
    data->fTipSize = tipSize;
    data->fSubd = subd;
    data->fRadius = radius;
    data->fXray = xray;
    data->fOperation = operation;

    MFloatVector vStart, vEnd, vFinal;
    switch (operation)
    {
    case 0:
        vEnd = vVector1;
        break;
    case 1:
        vFinal = vVector1 + vVector2;
        vEnd = vFinal;
        break;
    case 2:
        vFinal = vVector1 - vVector2;
        vEnd = vFinal;
        break;
    case 3:
        vFinal = vVector1 ^ vVector2;
        vEnd = vFinal;
        break;
    }

    vStart = MFloatVector();
    if (normalize)
        vEnd.normalize();

    data->fLineList.clear();
    DebugVector::drawArrow(vStart, vEnd, tipSize, radius, subd, data->fLineList);

    MGlobal::displayInfo("Draw prepared! Draw some shit right here dude!");

    return data;
}

void DebugVectorDrawOverride::addUIDrawables(const MDagPath& objPath,
                                             MHWRender::MUIDrawManager& drawManager,
                                             const MHWRender::MFrameContext& frameContext,
                                             const MUserData* data){
    /*
    Provides access to the MUIDrawManager, which can be used to queue up operations to draw simple UI
    shapes like lines, circles, text, etc.
    It is called after prepareForDraw() and carries the same restrictions on the sorts of operations it
    can perform.
        * objPath [MDagPath] is the path to the object being drawn.
        * drawManager [MUIDrawManager] it can be used to draw some simple geometry including text.
        * frameContext [MFrameContext] is the frame level context information.
        * data [MUserData] is the data cached by the prepareForDraw().
    */
    DebugVectorData* locatorData = (DebugVectorData*)data;
    if (!locatorData)
        MGlobal::displayInfo("Returning from addUIDrawables");
        return;

    MHWRender::DisplayStatus status = MHWRender::MGeometryUtilities::displayStatus(objPath);

    if (locatorData->fXray)
        drawManager.beginDrawInXray();
    else
        drawManager.beginDrawable();

    if (status == MHWRender::DisplayStatus::kActive)
        drawManager.setColor(locatorData->fActiveColor);
    else if (status == MHWRender::DisplayStatus::kLead)
        drawManager.setColor(locatorData->fLeadColor);
    else if (status == MHWRender::DisplayStatus::kDormant)
        drawManager.setColor(locatorData->fDormantColor);

    drawManager.setDepthPriority(5);

    drawManager.setLineWidth(locatorData->fLineWidth);
    drawManager.mesh(MHWRender::MUIDrawManager::kLines, locatorData->fLineList);

    if (locatorData->fXray)
        drawManager.endDrawInXray();
    else
        drawManager.endDrawable();

    MGlobal::displayInfo("Ending of drawing");

}